### ISSUE TITLE
Status code 200 when error in execution handler

### ISSUE THREAD

**Issue by eran-pinhas** [geopython/pywps#442]
# Description

Hey, 

I believe there's an issue with the status code. when the handler itself raises an error the Process.py catches it (lines below) sets the wps_request as failed (attached below) but then continues like no error occurred. As result the status code of the HTTP request is 200 and should be 500

https://github.com/geopython/pywps/blob/ec1d90f7424598a662f68c28e4de4452e2bdeaf3/pywps/app/Process.py#L232

I can think of 2 solutions:
* re-raise an NoApplicableCode at [Process.py#L232](https://github.com/geopython/pywps/blob/master/pywps/app/Process.py#L232)
* before the `return response` [(Service.py#L326)](https://github.com/geopython/pywps/blob/master/pywps/app/Service.py#326) check whether the response is failed one and if so raise an NoApplicableCode that will be catched at [Service.py#L335](https://github.com/geopython/pywps/blob/master/pywps/app/Service.py#L335)

What do you guys think? @cehbrecht @jachym 

# Environment

- operating system: ubuntu 18
- Python version: 3.6
- PyWPS version: latest from github
- source/distribution: git clone
- web server: Flask

**Comment by eran-pinhas** [comment]
@cehbrecht what do you think? I would be happy to help fixing the issue. just need your opinion on how to do it.

**Comment by cehbrecht** [comment]
We should use the Exceptions (HTTP status codes) according to the WPS protocol:
http://www.opengeospatial.org/standards/wps

For WPS 1.0.0 see Table 62 in the spec.

Here is a list of the WPS 2.0.0 protocol:
http://docs.opengeospatial.org/is/14-065/14-065.html

I'm not sure if we need to raise an exception when the process has failed. It will be reported with an status document ... and using status code=200. 

The exceptions available look like to report that there is something wrong with my WPS service. But to report process failures you just use a status document.

@jachym You are more familiar with the WPS protocol.

**Comment by cehbrecht** [comment]
Execute Exceptions in WPS 2.0.0:
http://docs.opengeospatial.org/is/14-065/14-065.html#61

**Comment by cehbrecht** [comment]
WPS 1.0.0:
> [..] If a process fails for some reason, the implementation raises an error and places it in the exception report included in this ProcessFailed structure.

It sounds confusing to me ... but the `ProcessFailed` structure is part of the status report ... not part of a WPS exception.

**Comment by jachym** [comment]
Hi,

this is confusing, as @cehbrecht says: ProcessFaild strucutre is part of status report - which returns 200 code.

But generally speaking, I would support more "aggressive" error handling and raise NoApplicableCode with 500 as response code.  

Re-raising in Process.py sounds like less invasive solution?

**Comment by eran-pinhas** [comment]
I agree about the ProcessFailed structure. but this is right only in the case of status request (which makes sense - the service have not failed at the status requesting). 

But I find both 1&2 protocols unclear in the case of synchronous request (status=false)

**Comment by cehbrecht** [comment]
@eran-pinhas ... `_run_process` is also called in async mode ... but it is a bit hidden:
https://github.com/geopython/pywps/blob/c351a11a4149d865308217497790d1f86991d02a/pywps/processing/job.py#L20

**Comment by gschwind** [comment]
Hello,

For the WPS 1.0.0 state section 10.3.1 :

> In the most primitive case, when a response form of RawDataOutput‖ is requested
> [...]
> In all other cases, the response to a valid Execute operation request is an
> ExecuteResponse XML document.

As far I understand is that if the request is valid, the response status is 200 in WPS 1.0.0. In WPS 2.0.0 it's seems different.

Best regards
